src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ChannelHandler.cs

```diff
@@ -615,6 +615,7 @@ private Task InitializeServiceChannel(ServiceChannel channel)
                 _serviceDispatcher.ChannelDispatcher.InitializeChannel((IClientChannel)channel.Proxy);
             }
 
+            // TODO: shouldn't this open be awaited? -> problem with lock outside
```

Comment:
I stumbled over this potentially bug/issue. I came to notice that calls to InitializeServiceChannel are NOT awaited:

https://github.com/CoreWCF/CoreWCF/blob/9bdd86153f04790cea42147da1e47ba3ede2f2a7/src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ChannelHandler.cs#L534
https://github.com/CoreWCF/CoreWCF/blob/9bdd86153f04790cea42147da1e47ba3ede2f2a7/src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/ChannelHandler.cs#L557

Awaits are not easily possible there because awaits are not supported within locks. And there are also out parameters which cannot be used with async-await. 

I would guess the channel creation could be reworked to use some SemaphoreSlim instead of a lock and the out parameters could be replaced with a named-tuple return. I didn't dare to touch this area because it seems very general overall CoreWCF and not specific to handling Duplex Channels, but I wanted to raise this concern.

Reply:
You're right, this is a potential race condition bug. I had a look and the code in GetSessionChannel is safe to replace with a SemaphoreSlim. This is the session creation code path, and the only other place ThisLock is used is in the session close code path. There isn't going to be any interaction between those two code paths, but even it's safe to replace the ThisLock with a SemaphoreSlim and use it in both locations.
The same is true of the _serviceDispatcher.ThisLock, it's only used in this one location so you can replace it with a SemaphoreSlim safely.

The problem you will hit is those methods use out parameters, so you will need to modify the method to use a Task of named tuple return value to be able to return multiple values. E.g.:
```c#
private async Task<(ServiceChannel serviceChannel, EndpointDispatcher endpoint, bool addressMatched)>  GetDatagramChannel(Message message)
```

Would you be able to fix this and make the async flow through correctly?

Reply:
Sure, I'll prepare the change later today,l and add it to this PR, should not take very long.